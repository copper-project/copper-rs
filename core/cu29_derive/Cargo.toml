[package]
name = "cu29-derive"
description = "This is the copper project runtime generator. It cannot be used independently from the copper project."
documentation = "https://docs.rs/cu29-derive"
version.workspace = true
authors.workspace = true
edition.workspace = true
license.workspace = true
keywords.workspace = true
categories.workspace = true
homepage.workspace = true
repository.workspace = true

[lib]
proc-macro = true

[dependencies]
cu29-runtime = { workspace = true }
cu29-traits = { workspace = true }
syn = { workspace = true }
quote = { workspace = true }
proc-macro2 = { workspace = true }
walkdir = "2.5.0"
syntect = "5.2.0"
itertools = "0.13.0"
convert_case = "0.6.0"

[build-dependencies]
cu29-unifiedlog = { workspace = true }  # needed

[dev-dependencies]
trybuild = "1.0"
cu29 = { workspace = true } # needed for compile_fail tests

[features]
default = []
# enables a more verbose build log showing the code generation.
macro_debug = []

