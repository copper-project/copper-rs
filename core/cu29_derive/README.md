## Copper main macros

This crate is part of the Copper project.
This allows you to invoke the main proc macro of Copper to generate your robot runtime. 

See the main crate cu29 for more information.