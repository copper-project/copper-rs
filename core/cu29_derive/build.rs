use std::path::{Path, PathBuf};
use std::{fs, io};

fn main() {
    println!("cargo::rerun-if-changed=tests/config");
    let trybuild_dir = PathBuf::from(std::env::var("OUT_DIR").unwrap())
        .ancestors()
        .nth(4)
        .unwrap()
        .to_path_buf()
        .join("tests/trybuild/cu29-derive/config");
    let config_dir =
        PathBuf::from(std::env::var("CARGO_MANIFEST_DIR").unwrap()).join("tests/config");
    copy_dir_all(config_dir, trybuild_dir).unwrap();
}

fn copy_dir_all(src: impl AsRef<Path>, dst: impl AsRef<Path>) -> io::Result<()> {
    fs::create_dir_all(&dst)?;
    for entry in fs::read_dir(src)? {
        let entry = entry?;
        let ty = entry.file_type()?;
        if ty.is_dir() {
            copy_dir_all(entry.path(), dst.as_ref().join(entry.file_name()))?;
        } else {
            fs::copy(entry.path(), dst.as_ref().join(entry.file_name()))?;
        }
    }
    Ok(())
}
