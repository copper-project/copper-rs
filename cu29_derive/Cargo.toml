[package]
name = "cu29-derive"
description = "This is the copper project runtime generator. It cannot be used independently from the copper project."
documentation = "https://docs.rs/cu29-derive"
version.workspace = true
authors.workspace = true
edition.workspace = true
license.workspace = true
keywords.workspace = true
categories.workspace = true
homepage.workspace = true
repository.workspace = true

[lib]
proc-macro = true

[dependencies]
cu29 = { workspace = true }
bincode = { workspace = true }
syn = { workspace = true }
quote = { workspace = true }
proc-macro2 = { workspace = true  }
walkdir = "2.5.0"
syntect = "5.2.0"
itertools = "0.13.0"

[build-dependencies]
cargo_metadata = "0.18.1"
cu29-unifiedlog = { workspace = true }

