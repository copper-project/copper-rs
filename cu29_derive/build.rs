use cargo_metadata::{MetadataCommand, Package};

fn main() {
    let metadata = MetadataCommand::new()
        .exec()
        .expect("Failed to fetch metadata");

    for package in metadata.packages {
        check_metadata(&package);
    }
}

fn check_metadata(package: &Package) {
    if let Some(metadata) = package.metadata.as_object() {
        if let Some(copper_plugin_type) = metadata.get("copper_plugin_type") {
            println!("cargo:warning=  --> Found copper-plugin-type in {}: {}", package.name, copper_plugin_type);
        }
    }
}

