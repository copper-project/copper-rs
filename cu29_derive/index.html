<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `cu29_derive` crate."><title>cu29_derive - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../static.files/rustdoc-46132b98.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="cu29_derive" data-themes="" data-resource-suffix="" data-rustdoc-version="1.86.0-nightly (48a426eca 2025-01-12)" data-channel="nightly" data-search-js="search-75f5ac3e.js" data-settings-js="settings-0f613d39.js" ><script src="../static.files/storage-59e33391.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-5f194d8c.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-044be391.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../cu29_derive/index.html">cu29_<wbr>derive</a><span class="version">0.5.2</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section id="rustdoc-toc"><h3><a href="#macros">Crate Items</a></h3><ul class="block"><li><a href="#macros" title="Macros">Macros</a></li><li><a href="#attributes" title="Attribute Macros">Attribute Macros</a></li></ul></section><div id="rustdoc-modnav"></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Crate <span>cu29_derive</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../src/cu29_derive/lib.rs.html#1-1433">Source</a> </span></div><h2 id="macros" class="section-header">Macros<a href="#macros" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="macro" href="macro.gen_cumsgs.html" title="macro cu29_derive::gen_cumsgs">gen_<wbr>cumsgs</a></div><div class="desc docblock-short">Generates the CopperList content type from a config.
gen_cumsgs!(“path/to/config.toml”)
It will create a new type called CuMsgs you can pass to the log reader for decoding:</div></li></ul><h2 id="attributes" class="section-header">Attribute Macros<a href="#attributes" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="attr" href="attr.copper_runtime.html" title="attr cu29_derive::copper_runtime">copper_<wbr>runtime</a></div><div class="desc docblock-short">Adds #[copper_runtime(config = “path”, sim_mode = false/true)] to your application struct to generate the runtime.
if sim_mode is omitted, it is set to false.
This will add a “runtime” field to your struct and implement the “new” and “run” methods.</div></li></ul></section></div></main></body></html>