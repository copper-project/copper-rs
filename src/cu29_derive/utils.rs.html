<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Source of the Rust file `core/cu29_derive/src/utils.rs`."><title>utils.rs - source</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-20739d33.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="cu29_derive" data-themes="" data-resource-suffix="" data-rustdoc-version="1.88.0-nightly (13e879094 2025-05-04)" data-channel="nightly" data-search-js="search-f7877310.js" data-settings-js="settings-5514c975.js" ><script src="../../static.files/storage-4e99c027.js"></script><script defer src="../../static.files/src-script-63605ae7.js"></script><script defer src="../../src-files.js"></script><script defer src="../../static.files/main-7ef8a74a.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc src"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="src-sidebar-title"><h2>Files</h2></div></nav><div class="sidebar-resizer"></div><main><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1><div class="sub-heading">cu29_derive/</div>utils.rs</h1><rustdoc-toolbar></rustdoc-toolbar></div><div class="example-wrap digits-3"><pre class="rust"><code><a href=#1 id=1 data-nosnippet>1</a><span class="kw">use </span>convert_case::{Case, Casing};
<a href=#2 id=2 data-nosnippet>2</a><span class="kw">use </span>std::path::PathBuf;
<a href=#3 id=3 data-nosnippet>3</a><span class="kw">use </span>walkdir::WalkDir;
<a href=#4 id=4 data-nosnippet>4</a>
<a href=#5 id=5 data-nosnippet>5</a><span class="doccomment">/// Small tool to create a valid enum entry from an identifier.
<a href=#6 id=6 data-nosnippet>6</a></span><span class="kw">pub</span>(<span class="kw">crate</span>) <span class="kw">fn </span>config_id_to_enum(id: <span class="kw-2">&amp;</span>str) -&gt; String {
<a href=#7 id=7 data-nosnippet>7</a>    <span class="kw">let </span><span class="kw-2">mut </span>candidate = id
<a href=#8 id=8 data-nosnippet>8</a>        .chars()
<a href=#9 id=9 data-nosnippet>9</a>        .map(|c| <span class="kw">if </span>c.is_alphanumeric() { c } <span class="kw">else </span>{ <span class="string">'_' </span>})
<a href=#10 id=10 data-nosnippet>10</a>        .collect::&lt;String&gt;();
<a href=#11 id=11 data-nosnippet>11</a>
<a href=#12 id=12 data-nosnippet>12</a>    candidate = candidate.to_case(Case::Pascal);
<a href=#13 id=13 data-nosnippet>13</a>
<a href=#14 id=14 data-nosnippet>14</a>    <span class="kw">if </span>candidate.chars().next().is_some_and(|c| c.is_ascii_digit()) {
<a href=#15 id=15 data-nosnippet>15</a>        candidate.insert(<span class="number">0</span>, <span class="string">'_'</span>);
<a href=#16 id=16 data-nosnippet>16</a>    }
<a href=#17 id=17 data-nosnippet>17</a>
<a href=#18 id=18 data-nosnippet>18</a>    candidate
<a href=#19 id=19 data-nosnippet>19</a>}
<a href=#20 id=20 data-nosnippet>20</a>
<a href=#21 id=21 data-nosnippet>21</a><span class="doccomment">/// Same as config_id_to_enum but for a struct member name
<a href=#22 id=22 data-nosnippet>22</a></span><span class="kw">pub</span>(<span class="kw">crate</span>) <span class="kw">fn </span>config_id_to_struct_member(id: <span class="kw-2">&amp;</span>str) -&gt; String {
<a href=#23 id=23 data-nosnippet>23</a>    <span class="kw">let </span><span class="kw-2">mut </span>candidate = id
<a href=#24 id=24 data-nosnippet>24</a>        .chars()
<a href=#25 id=25 data-nosnippet>25</a>        .map(|c| <span class="kw">if </span>c.is_alphanumeric() { c } <span class="kw">else </span>{ <span class="string">'_' </span>})
<a href=#26 id=26 data-nosnippet>26</a>        .collect::&lt;String&gt;();
<a href=#27 id=27 data-nosnippet>27</a>
<a href=#28 id=28 data-nosnippet>28</a>    candidate = candidate.to_case(Case::Snake);
<a href=#29 id=29 data-nosnippet>29</a>
<a href=#30 id=30 data-nosnippet>30</a>    <span class="kw">if </span>candidate.chars().next().is_some_and(|c| c.is_ascii_digit()) {
<a href=#31 id=31 data-nosnippet>31</a>        candidate.insert(<span class="number">0</span>, <span class="string">'_'</span>);
<a href=#32 id=32 data-nosnippet>32</a>    }
<a href=#33 id=33 data-nosnippet>33</a>
<a href=#34 id=34 data-nosnippet>34</a>    candidate
<a href=#35 id=35 data-nosnippet>35</a>}
<a href=#36 id=36 data-nosnippet>36</a>
<a href=#37 id=37 data-nosnippet>37</a><span class="comment">// Lifted this HORROR but it works.
<a href=#38 id=38 data-nosnippet>38</a></span><span class="kw">pub fn </span>caller_crate_root() -&gt; PathBuf {
<a href=#39 id=39 data-nosnippet>39</a>    <span class="kw">let </span>crate_name =
<a href=#40 id=40 data-nosnippet>40</a>        std::env::var(<span class="string">"CARGO_PKG_NAME"</span>).expect(<span class="string">"failed to read ENV var `CARGO_PKG_NAME`!"</span>);
<a href=#41 id=41 data-nosnippet>41</a>    <span class="kw">let </span>current_dir = std::env::current_dir().expect(<span class="string">"failed to unwrap env::current_dir()!"</span>);
<a href=#42 id=42 data-nosnippet>42</a>    <span class="kw">let </span>search_entry = <span class="macro">format!</span>(<span class="string">"name=\"{crate_name}\""</span>);
<a href=#43 id=43 data-nosnippet>43</a>    <span class="kw">for </span>entry <span class="kw">in </span>WalkDir::new(<span class="kw-2">&amp;</span>current_dir)
<a href=#44 id=44 data-nosnippet>44</a>        .into_iter()
<a href=#45 id=45 data-nosnippet>45</a>        .filter_entry(|e| !e.file_name().eq_ignore_ascii_case(<span class="string">"target"</span>))
<a href=#46 id=46 data-nosnippet>46</a>    {
<a href=#47 id=47 data-nosnippet>47</a>        <span class="kw">let </span><span class="prelude-val">Ok</span>(entry) = entry <span class="kw">else </span>{
<a href=#48 id=48 data-nosnippet>48</a>            <span class="kw">continue</span>;
<a href=#49 id=49 data-nosnippet>49</a>        };
<a href=#50 id=50 data-nosnippet>50</a>        <span class="kw">if </span>!entry.file_type().is_file() {
<a href=#51 id=51 data-nosnippet>51</a>            <span class="kw">continue</span>;
<a href=#52 id=52 data-nosnippet>52</a>        }
<a href=#53 id=53 data-nosnippet>53</a>        <span class="kw">let </span><span class="prelude-val">Some</span>(file_name) = entry.path().file_name() <span class="kw">else </span>{
<a href=#54 id=54 data-nosnippet>54</a>            <span class="kw">continue</span>;
<a href=#55 id=55 data-nosnippet>55</a>        };
<a href=#56 id=56 data-nosnippet>56</a>        <span class="kw">if </span>!file_name.eq_ignore_ascii_case(<span class="string">"Cargo.toml"</span>) {
<a href=#57 id=57 data-nosnippet>57</a>            <span class="kw">continue</span>;
<a href=#58 id=58 data-nosnippet>58</a>        }
<a href=#59 id=59 data-nosnippet>59</a>        <span class="kw">let </span><span class="prelude-val">Ok</span>(cargo_toml) = std::fs::read_to_string(entry.path()) <span class="kw">else </span>{
<a href=#60 id=60 data-nosnippet>60</a>            <span class="kw">continue</span>;
<a href=#61 id=61 data-nosnippet>61</a>        };
<a href=#62 id=62 data-nosnippet>62</a>        <span class="kw">if </span>cargo_toml
<a href=#63 id=63 data-nosnippet>63</a>            .chars()
<a href=#64 id=64 data-nosnippet>64</a>            .filter(|<span class="kw-2">&amp;</span>c| !c.is_whitespace())
<a href=#65 id=65 data-nosnippet>65</a>            .collect::&lt;String&gt;()
<a href=#66 id=66 data-nosnippet>66</a>            .contains(search_entry.as_str())
<a href=#67 id=67 data-nosnippet>67</a>        {
<a href=#68 id=68 data-nosnippet>68</a>            <span class="kw">return </span>entry.path().parent().unwrap().to_path_buf();
<a href=#69 id=69 data-nosnippet>69</a>        }
<a href=#70 id=70 data-nosnippet>70</a>    }
<a href=#71 id=71 data-nosnippet>71</a>    current_dir
<a href=#72 id=72 data-nosnippet>72</a>}
<a href=#73 id=73 data-nosnippet>73</a>
<a href=#74 id=74 data-nosnippet>74</a><span class="attr">#[cfg(test)]
<a href=#75 id=75 data-nosnippet>75</a></span><span class="kw">mod </span>tests {
<a href=#76 id=76 data-nosnippet>76</a>    <span class="kw">use </span><span class="kw">crate</span>::utils::config_id_to_enum;
<a href=#77 id=77 data-nosnippet>77</a>
<a href=#78 id=78 data-nosnippet>78</a>    <span class="kw">fn </span>is_valid_rust_identifier(input: <span class="kw-2">&amp;</span>str) -&gt; bool {
<a href=#79 id=79 data-nosnippet>79</a>        <span class="kw">if </span>input.is_empty() {
<a href=#80 id=80 data-nosnippet>80</a>            <span class="kw">return </span><span class="bool-val">false</span>;
<a href=#81 id=81 data-nosnippet>81</a>        }
<a href=#82 id=82 data-nosnippet>82</a>
<a href=#83 id=83 data-nosnippet>83</a>        <span class="comment">// Check if the first character is valid
<a href=#84 id=84 data-nosnippet>84</a>        </span><span class="kw">let </span><span class="kw-2">mut </span>chars = input.chars();
<a href=#85 id=85 data-nosnippet>85</a>        <span class="kw">if let </span><span class="prelude-val">Some</span>(first) = chars.next() {
<a href=#86 id=86 data-nosnippet>86</a>            <span class="kw">if </span>!first.is_alphabetic() &amp;&amp; first != <span class="string">'_' </span>{
<a href=#87 id=87 data-nosnippet>87</a>                <span class="kw">return </span><span class="bool-val">false</span>;
<a href=#88 id=88 data-nosnippet>88</a>            }
<a href=#89 id=89 data-nosnippet>89</a>        }
<a href=#90 id=90 data-nosnippet>90</a>
<a href=#91 id=91 data-nosnippet>91</a>        <span class="comment">// Check the rest of the characters
<a href=#92 id=92 data-nosnippet>92</a>        </span><span class="kw">if </span>!chars.all(|c| c.is_alphanumeric() || c == <span class="string">'_'</span>) {
<a href=#93 id=93 data-nosnippet>93</a>            <span class="kw">return </span><span class="bool-val">false</span>;
<a href=#94 id=94 data-nosnippet>94</a>        }
<a href=#95 id=95 data-nosnippet>95</a>
<a href=#96 id=96 data-nosnippet>96</a>        <span class="comment">// Check if it's a Rust keyword (can use a set of known keywords)
<a href=#97 id=97 data-nosnippet>97</a>        </span><span class="kw">let </span>keywords = [
<a href=#98 id=98 data-nosnippet>98</a>            <span class="string">"as"</span>, <span class="string">"break"</span>, <span class="string">"const"</span>, <span class="string">"continue"</span>, <span class="string">"crate"</span>, <span class="string">"else"</span>, <span class="string">"enum"</span>, <span class="string">"extern"</span>, <span class="string">"false"</span>, <span class="string">"fn"</span>,
<a href=#99 id=99 data-nosnippet>99</a>            <span class="string">"for"</span>, <span class="string">"if"</span>, <span class="string">"impl"</span>, <span class="string">"in"</span>, <span class="string">"let"</span>, <span class="string">"loop"</span>, <span class="string">"match"</span>, <span class="string">"mod"</span>, <span class="string">"move"</span>, <span class="string">"mut"</span>, <span class="string">"pub"</span>, <span class="string">"ref"</span>,
<a href=#100 id=100 data-nosnippet>100</a>            <span class="string">"return"</span>, <span class="string">"self"</span>, <span class="string">"Self"</span>, <span class="string">"static"</span>, <span class="string">"struct"</span>, <span class="string">"super"</span>, <span class="string">"trait"</span>, <span class="string">"true"</span>, <span class="string">"type"</span>,
<a href=#101 id=101 data-nosnippet>101</a>            <span class="string">"unsafe"</span>, <span class="string">"use"</span>, <span class="string">"where"</span>, <span class="string">"while"</span>,
<a href=#102 id=102 data-nosnippet>102</a>        ];
<a href=#103 id=103 data-nosnippet>103</a>
<a href=#104 id=104 data-nosnippet>104</a>        !keywords.contains(<span class="kw-2">&amp;</span>input)
<a href=#105 id=105 data-nosnippet>105</a>    }
<a href=#106 id=106 data-nosnippet>106</a>
<a href=#107 id=107 data-nosnippet>107</a>    <span class="attr">#[test]
<a href=#108 id=108 data-nosnippet>108</a>    </span><span class="kw">fn </span>test_identifier_to_enum() {
<a href=#109 id=109 data-nosnippet>109</a>        <span class="kw">let </span>test_cases = [<span class="string">"toto"</span>, <span class="string">"#id"</span>, <span class="string">"!!something"</span>, <span class="string">"hey?"</span>, <span class="string">"é"</span>, <span class="string">"t"</span>];
<a href=#110 id=110 data-nosnippet>110</a>
<a href=#111 id=111 data-nosnippet>111</a>        test_cases.iter().for_each(|input| {
<a href=#112 id=112 data-nosnippet>112</a>            <span class="kw">let </span>after = config_id_to_enum(input);
<a href=#113 id=113 data-nosnippet>113</a>            <span class="macro">assert!</span>(
<a href=#114 id=114 data-nosnippet>114</a>                is_valid_rust_identifier(after.as_str()),
<a href=#115 id=115 data-nosnippet>115</a>                <span class="string">"bf {input} af {after}"
<a href=#116 id=116 data-nosnippet>116</a>            </span>);
<a href=#117 id=117 data-nosnippet>117</a>        })
<a href=#118 id=118 data-nosnippet>118</a>    }
<a href=#119 id=119 data-nosnippet>119</a>
<a href=#120 id=120 data-nosnippet>120</a>    <span class="attr">#[test]
<a href=#121 id=121 data-nosnippet>121</a>    </span><span class="kw">fn </span>test_identifier_to_struct_member() {
<a href=#122 id=122 data-nosnippet>122</a>        <span class="macro">assert_eq!</span>(<span class="kw">crate</span>::utils::config_id_to_struct_member(<span class="string">"toto"</span>), <span class="string">"toto"</span>);
<a href=#123 id=123 data-nosnippet>123</a>        <span class="macro">assert_eq!</span>(<span class="kw">crate</span>::utils::config_id_to_struct_member(<span class="string">"#id"</span>), <span class="string">"id"</span>);
<a href=#124 id=124 data-nosnippet>124</a>        <span class="macro">assert_eq!</span>(
<a href=#125 id=125 data-nosnippet>125</a>            <span class="kw">crate</span>::utils::config_id_to_struct_member(<span class="string">"!!something"</span>),
<a href=#126 id=126 data-nosnippet>126</a>            <span class="string">"something"
<a href=#127 id=127 data-nosnippet>127</a>        </span>);
<a href=#128 id=128 data-nosnippet>128</a>        <span class="macro">assert_eq!</span>(<span class="kw">crate</span>::utils::config_id_to_struct_member(<span class="string">"hey?"</span>), <span class="string">"hey"</span>);
<a href=#129 id=129 data-nosnippet>129</a>        <span class="macro">assert_eq!</span>(<span class="kw">crate</span>::utils::config_id_to_struct_member(<span class="string">"é"</span>), <span class="string">"é"</span>);
<a href=#130 id=130 data-nosnippet>130</a>        <span class="macro">assert_eq!</span>(<span class="kw">crate</span>::utils::config_id_to_struct_member(<span class="string">"T"</span>), <span class="string">"t"</span>);
<a href=#131 id=131 data-nosnippet>131</a>        <span class="macro">assert_eq!</span>(
<a href=#132 id=132 data-nosnippet>132</a>            <span class="kw">crate</span>::utils::config_id_to_struct_member(<span class="string">"Test_Dunder"</span>),
<a href=#133 id=133 data-nosnippet>133</a>            <span class="string">"test_dunder"
<a href=#134 id=134 data-nosnippet>134</a>        </span>);
<a href=#135 id=135 data-nosnippet>135</a>    }
<a href=#136 id=136 data-nosnippet>136</a>}</code></pre></div></section></main></body></html>