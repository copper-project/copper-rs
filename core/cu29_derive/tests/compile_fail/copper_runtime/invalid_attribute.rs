use cu29_derive::copper_runtime;

const CONFIG_FILE: &str = "/path/to/config.ron";

#[copper_runtime(config = CONFIG_FILE)]
struct MyApplicationStruct;

fn main() {}