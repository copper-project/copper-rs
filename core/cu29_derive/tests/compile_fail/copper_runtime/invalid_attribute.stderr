error: expected string literal
 --> tests/compile_fail/copper_runtime/invalid_attribute.rs:5:27
  |
5 | #[copper_runtime(config = CONFIG_FILE)]
  |                           ^^^^^^^^^^^
